File: service/internal/s3shared/s3100continue.go
package s3shared

import (
	"context"
	"fmt"
	"github.com/aws/smithy-go/middleware"
	smithyhttp "github.com/aws/smithy-go/transport/http"
)

const s3100ContinueID = "S3100Continue"
const defaultLimit int64 = 1024 * 1024 * 2

// Add100Continue add middleware, which adds {Expect: 100-continue} header for s3 client HTTP PUT request larger than 2MB
// or with unknown size streaming bodies, during operation builder step
func Add100Continue(stack *middleware.Stack, option AddContinueOption) error {
	return stack.Build.Add(&s3100Continue{
		continueHeaderThresholdBytes: option.ContinueHeaderThresholdBytes,
	}, middleware.After)
}

type s3100Continue struct {
	continueHeaderThresholdBytes int64
}

// ID returns the middleware identifier
func (m *s3100Continue) ID() string {
	return s3100ContinueID
}

func (m *s3100Continue) HandleBuild(
	ctx context.Context, in middleware.BuildInput, next middleware.BuildHandler,
) (
	out middleware.BuildOutput, metadata middleware.Metadata, err error,
) {
	sizeLimit := defaultLimit
	switch {
	case m.continueHeaderThresholdBytes == -1:
		return next.HandleBuild(ctx, in)
	case m.continueHeaderThresholdBytes > 0:
		sizeLimit = m.continueHeaderThresholdBytes
	default:
	}

	req, ok := in.Request.(*smithyhttp.Request)
	if !ok {
		return out, metadata, fmt.Errorf("unknown request type %T", req)
	}

	if req.ContentLength == -1 || (req.ContentLength == 0 && req.Body != nil) || req.ContentLength >= sizeLimit {
		req.Header.Set("Expect", "100-continue")
	}

	return next.HandleBuild(ctx, in)
}

// AddContinueOption passes user configuration of threshold size to trigger 100-continue header
type AddContinueOption struct {
	ContinueHeaderThresholdBytes int64
}

File: codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/S3100Continue.java
package software.amazon.smithy.aws.go.codegen.customization;

import software.amazon.smithy.codegen.core.SymbolProvider;
import software.amazon.smithy.go.codegen.GoDelegator;
import software.amazon.smithy.go.codegen.GoSettings;
import software.amazon.smithy.go.codegen.GoWriter;
import software.amazon.smithy.go.codegen.SymbolUtils;
import software.amazon.smithy.go.codegen.integration.ConfigField;
import software.amazon.smithy.go.codegen.integration.GoIntegration;
import software.amazon.smithy.go.codegen.integration.MiddlewareRegistrar;
import software.amazon.smithy.go.codegen.integration.RuntimeClientPlugin;
import software.amazon.smithy.model.Model;
import software.amazon.smithy.model.shapes.ServiceShape;
import software.amazon.smithy.model.traits.HttpTrait;
import software.amazon.smithy.utils.ListUtils;

import java.util.List;

/**
 * Add middleware, which adds {Expect: 100-continue} header for s3 client HTTP PUT request larger than 2MB
 * or with unknown size streaming bodies, during operation builder step
 */
public class S3100Continue implements GoIntegration {
    private static final String ADD_100Continue_Header = "add100Continue";
    private static final String ADD_100Continue_Header_INTERNAL = "Add100Continue";
    private static final String ADD_100Continue_Header_Option = "AddContinueOption";
    private static final String Continue_Client_Option = "ContinueHeaderThresholdBytes";

    /**
     * Return true if service is Amazon S3.
     *
     * @param model   is the generation model.
     * @param service is the service shape being audited.
     */
    private static boolean isS3Service(Model model, ServiceShape service) {
        return S3ModelUtils.isServiceS3(model, service);
    }

    /**
     * Gets the sort order of the customization from -128 to 127, with lowest
     * executed first.
     *
     * @return Returns the sort order, defaults to -40.
     */
    @Override
    public byte getOrder() {
        return 126;
    }

    @Override
    public void writeAdditionalFiles(
            GoSettings settings,
            Model model,
            SymbolProvider symbolProvider,
            GoDelegator goDelegator
    ) {
        ServiceShape service = settings.getService(model);
        if (!isS3Service(model, service)) {
            return;
        }

        goDelegator.useShapeWriter(service, this::writeMiddlewareHelper);
    }

    private void writeMiddlewareHelper(GoWriter writer) {
        writer.openBlock("func $L(stack *middleware.Stack, options Options) error {", "}", ADD_100Continue_Header, () -> {
            writer.openBlock("return $T(stack, $T{", "})",
                    SymbolUtils.createValueSymbolBuilder(ADD_100Continue_Header_INTERNAL,
                            AwsCustomGoDependency.S3_SHARED_CUSTOMIZATION).build(),
                    SymbolUtils.createValueSymbolBuilder(ADD_100Continue_Header_Option,
                            AwsCustomGoDependency.S3_SHARED_CUSTOMIZATION).build(), () -> {
                        writer.write("ContinueHeaderThresholdBytes: options.ContinueHeaderThresholdBytes,"
                        );
                    }
                    );
        });
        writer.insertTrailingNewline();
    }

    @Override
    public List<RuntimeClientPlugin> getClientPlugins() {
        return ListUtils.of(
                RuntimeClientPlugin.builder()
                        .operationPredicate((model, service, operation) -> isS3Service(model, service) && operation.
                                getTrait(HttpTrait.class).get().getMethod().equals("PUT"))
                        .registerMiddleware(MiddlewareRegistrar.builder()
                                .resolvedFunction(SymbolUtils.createValueSymbolBuilder(ADD_100Continue_Header).build())
                                .useClientOptions()
                                .build()
                        )
                        .build(),
                RuntimeClientPlugin.builder()
                        .servicePredicate(S3100Continue::isS3Service)
                        .configFields(ListUtils.of(
                            ConfigField.builder()
                                    .name(Continue_Client_Option)
                                    .type(SymbolUtils.createValueSymbolBuilder("int64")
                                            .putProperty(SymbolUtils.GO_UNIVERSE_TYPE, true)
                                            .build())
                                    .documentation("The threshold ContentLength for HTTP PUT request to receive {Expect: 100-continue} header. " +
                                            "When set to -1, this header will be opt out of the operation request; when set to 0, the threshold" +
                                            "will be set to default 2MB")
                                    .build()
                        ))
                        .build()
        );
    }
}
